carapace-server/src/main/java/org/carapaceproxy/server/certificates/ocsp/OcspStaplingManager.java

```diff
@@ -0,0 +1,194 @@
+        if (period <= 0) {
+            return;
+        }
+        if (scheduler == null) {
+            scheduler = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat(THREAD_NAME).build());
+        }
+
+        //LOG.info("Starting " + OcspStaplingManager.class.getName() + ", period: " + period + " seconds" + (TESTING_MODE ? " (TESTING_MODE)" : ""));
+        scheduledFuture = scheduler.scheduleWithFixedDelay(this, 0, period, TimeUnit.SECONDS);
+    }
+
+    public synchronized void stop() {
+        started = false;
+        if (scheduler != null) {
+            scheduler.shutdown();
+            try {
+                scheduler.awaitTermination(10, TimeUnit.SECONDS);
+                scheduler = null;
+                scheduledFuture = null;
+            } catch (InterruptedException err) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    @Override
+    public void run() {
+        if (groupMembershipHandler != null) {
+            int acquirePeriod = period;
+            if (acquirePeriod <= 0) {
+                acquirePeriod = 1;
+            }
+            groupMembershipHandler.executeInMutex(THREAD_NAME, acquirePeriod, () -> {
+                // Only one node of the cluster executes this
+                ocspChecks.entrySet().forEach((ocspCheck) -> {
+                    String domain = ocspCheck.getKey();
+                    OcspCheck check = ocspCheck.getValue();
+                    try {
+                        if (isExpired(check.response)) {
+                            performStaplingForDomain(domain, check.chain);
+                        }
+                    } catch (IOException | OCSPException | GeneralSecurityException ex) {
+                        LOG.log(Level.SEVERE, "OCSP stapling failed for domain " + domain);
+                    }
+                });
+            });
+        }
+    }
+
+    public static boolean isExpired(OCSPResp response) throws OCSPException {
+        BasicOCSPResp basicResponse = (BasicOCSPResp) response.getResponseObject();
+        Date nextUpdate = basicResponse.getResponses()[0].getNextUpdate();
+        return nextUpdate == null || nextUpdate.before(new Date());
+    }
+
+    public void performStaplingForDomain(String domain, Certificate[] chain) throws IOException, OCSPException, GeneralSecurityException {
+        if (!OpenSsl.isAvailable() || !OpenSsl.isOcspSupported() || chain == null && chain.length == 0) {
+            LOG.log(Level.INFO, "OCSP non supported. skipping OCSP stapling");
+            return;
+        }
+        X509Certificate cert = (X509Certificate) chain[0];
+        X509Certificate issuer = (X509Certificate) chain[chain.length - 1];
+
+        URI uri = OcspUtils.ocspUri(cert);
+        LOG.log(Level.INFO, "OCSP Responder URI: " + uri);
+
+        if (uri == null) {
+            LOG.log(Level.INFO, "The CA/certificate doesn't have an OCSP responder, skipping OCSP stapling");
```

Comment:
Please report something about domain or anything available to make this line useful when you have thouthands of certificates